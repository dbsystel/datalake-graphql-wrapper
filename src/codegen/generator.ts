/*
 * SPDX-FileCopyrightText: 2023 DB Systel GmbH
 * SPDX-License-Identifier: Apache-2.0
 */
import dotenv from 'dotenv'
import { camelCase, upperFirst } from 'lodash-es'
import * as path from 'path'
import * as fs from 'fs'
import * as ejs from 'ejs'
import { SqlClient } from 'src/client/sql-client'
import replaceSpecialCharacters from 'replace-special-characters'
import {
  parseFields,
  generateJsonSchema,
  generateModel,
} from 'src/helpers/convert'
import { globby } from 'globby'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

dotenv.config()

type SchemaTableColumn = { name: string; type: string; description: string }

export const prepare = async (schema: string, tableName: string) => {
  const importPath = `src/schema/automated/${process.env.ENGINE_CATALOG}/${schema}/${tableName}`

  const targetPath = path.join(
    path.dirname(__dirname),
    '/schema/automated',
    `${process.env.ENGINE_CATALOG}/`,
    `${schema}/`,
    `${tableName}/`,
  )

  if (fs.existsSync(targetPath)) {
    fs.rmSync(targetPath, { recursive: true, force: true })
  }

  fs.mkdirSync(targetPath, { recursive: true })

  const baseClassName = `${upperFirst(
    camelCase(process.env.ENGINE_CATALOG),
  )}_${upperFirst(camelCase(schema))}_${upperFirst(camelCase(tableName))}`

  const queryName =
    `${process.env.ENGINE_CATALOG}_${schema}_${tableName}`.toLowerCase()

  const tableColumns = await getSchemaTableColumns(schema, tableName)
  const parsedFields = parseFields(tableColumns)
  const jsonSchema = generateJsonSchema({ source: parsedFields })
  const models = generateModel({
    source: jsonSchema,
    name: baseClassName,
    models: {},
    isRoot: true,
  })

  //@ts-ignore
  const mainModel = Object.values(models).find((ele) => ele.root == true)
  //@ts-ignore
  const filterFields = Object.values(mainModel.fields).filter(
    //@ts-ignore
    (ele) => ele.filter == true,
  )

  //@ts-ignore
  const dateFields = Object.entries(jsonSchema.properties)
    //@ts-ignore
    .filter((ele) => ele[1].type == 'date-time')
    .map((ele) => ele[0])

  return {
    schema,
    tableName,
    importPath,
    targetPath,
    baseClassName,
    queryName,
    tableColumns,
    parsedFields,
    jsonSchema,
    jsonSchemaAsString: JSON.stringify(jsonSchema, null, 2),
    models,
    filterFields,
    dateFields,
  }
}

const getSchemaTableColumns = async (
  schema: string,
  tableName: string,
): Promise<SchemaTableColumn[]> => {
  const client = new SqlClient()

  const rawTableColumns = await client.query({
    query: `SHOW COLUMNS from "${process.env.ENGINE_CATALOG}"."${schema}"."${tableName}"`,
    schema,
  })

  /**
   * since there could be special characters in the column name
   * we have to replace them
   *
   * INFO: This is not a problem on our side, it's a problem from the source
   * and they should update their schema to have valid column names
   */
  const tableColumns: SchemaTableColumn[] = rawTableColumns.map((ele) => {
    return {
      name: replaceSpecialCharacters(ele.Column),
      type: ele.Type,
      description: '',
    }
  })

  return tableColumns
}

export const generateIndexFile = async () => {
  const foundSchema = await globby('src/schema', {
    expandDirectories: {
      files: ['schema'],
      extensions: ['ts'],
    },
  })

  const indexFilePath = path.join(path.dirname(__dirname), '/schema/')
  if (fs.existsSync(`${path.join(indexFilePath, '/index.ts')}`)) {
    fs.rmSync(`${path.join(indexFilePath, '/index.ts')}`)
  }

  generateFromTemplate('index', { foundSchema }, indexFilePath, 'index.ts')
}

export const generateFromTemplate = (
  template: string,
  templateOptions: { [key: string]: any },
  targetPath: string,
  targetFilename: string,
) => {
  const templateFile = path.join(
    path.dirname(__dirname),
    `/codegen/templates/${template}.ejs.t`,
  )

  if (!fs.existsSync(templateFile)) {
    throw 'Template file does not exist: ' + templateFile
  }

  fs.readFile(templateFile, 'utf8', function (err, data) {
    if (err) throw err

    let generatedContent = ''

    generatedContent = ejs.render(data, templateOptions, {
      compileDebug: true,
    })

    fs.writeFileSync(path.join(targetPath, targetFilename), generatedContent)
  })
}
