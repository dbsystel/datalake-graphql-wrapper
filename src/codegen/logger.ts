/*
 * SPDX-FileCopyrightText: 2022 Blitz Revolution Inc.
 * SPDX-License-Identifier: MIT
 *
 * Source: https://github.com/blitz-js/blitz/blob/main/packages/blitz/src/logging.ts
 */
import c from 'chalk'
import ora from 'ora'
import readline from 'readline'

export const newLine = () => {
  console.log(' ')
}

export const chalk = c

export const withCaret = (str: string) => {
  return `${c.gray('>')} ${str}`
}

export const withCheck = (str: string) => {
  return `${c.green('âœ”')} ${str}`
}

export const withProgress = (str: string) => {
  return withCaret(str)
}

export const greenText = (str: string) => {
  return `${c.green(str)}`
}

/**
 * Clears the line and optionally log a message to stdout.
 *
 * @param {string} msg
 */
export const clearLine = (msg?: string) => {
  readline.clearLine(process.stdout, 0)
  readline.cursorTo(process.stdout, 0)
  msg && process.stdout.write(msg)
}

export const clearConsole = () => {
  if (process.platform === 'win32') {
    process.stdout.write('\x1B[2J\x1B[0f')
  } else {
    process.stdout.write('\x1B[2J\x1B[3J\x1B[H')
  }
}

/**
 * Logs a progress message to stdout.
 *
 * @param {string} msg
 */
export const progress = (msg: string) => {
  console.log(withProgress(msg))
}

export const spinner = (str?: string) => {
  return ora({
    text: str,
  })
}

/**
 * Logs a green success message to stdout.
 *
 * @param {string} msg
 */
export const success = (msg: string) => {
  console.log(withCheck(c.green(msg)))
}

/**
 * Logs a red error message to stdout.
 *
 * @param {string} msg
 */
export const error = (msg: string) => {
  console.log(`${c.red(msg)}`)
}

/**
 * Colorizes a variable for display.
 *
 * @param {string} val
 */
export const variable = (val: any) => {
  return c.cyan.bold(`${val}`)
}
