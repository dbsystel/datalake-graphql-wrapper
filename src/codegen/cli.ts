/*
 * SPDX-FileCopyrightText: 2023 DB Systel GmbH
 * SPDX-License-Identifier: Apache-2.0
 */
import dotenv from 'dotenv'
import prompts, { Choice } from 'prompts'
import {
  generateFromTemplate,
  generateIndexFile,
  prepare,
} from 'src/codegen/generator'
import replaceSpecialCharacters from 'replace-special-characters'
import { SqlClient } from 'src/client/sql-client'
import { spinner, withCheck } from 'src/codegen/logger'
import c from 'chalk'
import { spawn } from 'cross-spawn'
import args from 'args'

dotenv.config()

args
  .option('schema', 'schema to use')
  .option('table', 'table to use')
  .option('index', 'Generate index file')
  .option('onlyindex', '(Re-)Generate the index file')
  .option('lint', 'Lint the generated code')

const flags = args.parse(process.argv)

;(async () => {
  const onState = (state: { value: any; aborted: boolean }) => {
    if (state.aborted) {
      process.nextTick(() => {
        process.exit(0)
      })
    }
  }

  if (
    flags.onlyindex &&
    (flags.onlyindex === 'true' || flags.onlyindex === true)
  ) {
    await generateIndexFile()
    console.log(withCheck(c.bold('Index file with all imports generated!')))
    await lintCode('src/schema/index.ts')
    console.log(withCheck(c.bold('Linted!')))
  } else {
    const client = new SqlClient()

    let schemaResponse
    let selectedTables

    if (!flags.schema || typeof flags.schema !== 'string') {
      const fetchRemoteSchemas = await client
        .query({
          query: `SHOW SCHEMAS from ${process.env.ENGINE_CATALOG}`,
        })
        .catch(() => {
          process.exit(1)
        })

      const schemas: Choice[] = fetchRemoteSchemas.map((ele) => {
        return { title: ele.Schema, value: ele.Schema }
      })

      schemaResponse = await prompts({
        type: 'autocomplete',
        name: 'schema',
        message: 'Choose a schema',
        choices: schemas,
        onState,
      })
    }

    const schemaResult =
      flags.schema && typeof flags.schema == 'string'
        ? flags.schema
        : schemaResponse?.schema

    if (!flags.table) {
      const fetchRemoteTables = await client
        .query({
          query: `SHOW TABLES from "${process.env.ENGINE_CATALOG}"."${schemaResult}"`,
        })
        .catch(() => {
          process.exit(1)
        })

      const tableNames: Choice[] = fetchRemoteTables.map((ele) => {
        return {
          title: ele.Table,
          value: replaceSpecialCharacters(ele.Table),
        }
      })

      tableNames.unshift({ title: '### All tables ###', value: 'all' })

      const tableResponse: { tables: string[] } = await prompts({
        type: 'autocompleteMultiselect',
        name: 'tables',
        message: 'Choose some tables',
        choices: tableNames,
        initial: 0,
        onState,
      })

      selectedTables = tableResponse.tables.includes('all')
        ? tableNames.map((ele) => ele.value).filter((ele) => ele != 'all')
        : tableResponse.tables
    } else {
      selectedTables =
        typeof flags.table == 'string' ? [flags.table] : flags.table
    }
    const progressSpinner = spinner().start()

    for (const tableName of selectedTables) {
      progressSpinner.text = `Generate code for table ${tableName}...`
      const preparedTable = await prepare(schemaResult, tableName)
      generateFromTemplate(
        'interface',
        preparedTable,
        preparedTable.targetPath,
        'interface.ts',
      )

      generateFromTemplate(
        'schema',
        preparedTable,
        preparedTable.targetPath,
        'schema.ts',
      )
      generateFromTemplate(
        'jsonSchema',
        preparedTable,
        preparedTable.targetPath,
        'json-schema.ts',
      )
    }
    progressSpinner.succeed(c.bold('All files generated!'))

    let indexResponse

    if (flags.index == null || flags.index == undefined) {
      indexResponse = await prompts({
        type: 'toggle',
        name: 'index',
        message: 'Generate index file?',
        initial: true,
        onState,
      })
    }

    const shouldIndex =
      flags.index == true || flags.index == 'true' || indexResponse?.index

    if (shouldIndex) {
      await generateIndexFile()
      console.log(withCheck(c.bold('Index file with all imports generated!')))
    }

    let lintResponse
    if (flags.lint == null || flags.lint == undefined) {
      lintResponse = await prompts({
        type: 'toggle',
        name: 'lint',
        message: 'Run linter?',
        initial: true,
        onState,
      })
    }

    const shouldLint =
      flags.lint == true || flags.lint == 'true' || lintResponse?.lint

    if (shouldLint) {
      await lintCode()
      console.log(withCheck(c.bold('Linted!')))
    }
    console.log(withCheck(c.bold(`Done! Let's gooooooo...`)))
  }
})()

async function lintCode(path = 'src/schema/**/*.*') {
  await new Promise((resolve) => {
    const cp = spawn(
      `eslint`,
      [path, '--fix', '--config', '.eslintrc.lint.cjs'],
      {
        stdio: ['inherit', 'pipe', 'pipe'],
      },
    )
    cp.on('exit', resolve)
  })
}
