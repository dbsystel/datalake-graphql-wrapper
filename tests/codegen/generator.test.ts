import { expect, test, describe, vi } from 'vitest'
import { getSchemaTableColumns } from '../../src/codegen/generator'
import { sqlTableColumns } from '../mockdata'

describe('Codegen - Generator', () => {
  test('getSchemaTableColumns', async () => {
    const fnResult = await getSchemaTableColumns('testSchema', 'testTable')
    expect(fnResult).toMatchSnapshot()
  })
})

//Mock the Presto Client
vi.mock('src/client/sql-client', () => {
  const SqlClient = vi.fn()
  SqlClient.prototype.query = vi.fn(async () => {
    return sqlTableColumns
  })
  SqlClient.prototype.end = vi.fn()

  return { SqlClient }
})
