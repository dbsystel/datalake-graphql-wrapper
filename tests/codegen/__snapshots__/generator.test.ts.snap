// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Codegen - Generator > getSchemaTableColumns 1`] = `
[
  {
    "description": "",
    "name": "stringarray",
    "type": "array(varchar)",
  },
  {
    "description": "",
    "name": "stringfield",
    "type": "varchar",
  },
  {
    "description": "",
    "name": "booleanfield",
    "type": "boolean",
  },
  {
    "description": "",
    "name": "datetimefield",
    "type": "timestamp(3)",
  },
  {
    "description": "",
    "name": "bigintfield",
    "type": "bigint",
  },
  {
    "description": "",
    "name": "simpleobject",
    "type": "row(sosub1 varchar, sosub2 bigint)",
  },
  {
    "description": "",
    "name": "complexobject",
    "type": "row(cosub1 varchar, cosub2 array(varchar), cosub3 array(row(cosub31 varchar, cosub32 varchar)))",
  },
  {
    "description": "",
    "name": "dash-field",
    "type": "varchar",
  },
  {
    "description": "",
    "name": "special-aou",
    "type": "varchar",
  },
]
`;
